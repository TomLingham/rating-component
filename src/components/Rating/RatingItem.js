// @flow

import * as React from 'react';
import styled, { css } from 'styled-components';

import { type RatingIconType } from '../RatingIcons';

const zoomMixin = css`
  svg {
    transform: scale(1.2);
  }
`;

const Container = styled.span`
  cursor: pointer;
  padding: 0 4px;
  width: 100%;
  transition: zoom 180ms, opacity 180ms;
  opacity: ${p => (p.selected || p.active ? 1 : 0.5)};
  ${p => p.active && zoomMixin};
`;

type Props = {
  color: string,
  icon: RatingIconType,
  active: boolean,
  selected: boolean,
  fillPercentage: number,
};

export default ({
  color = '#fff',
  icon: Icon,
  fillPercentage,
  ...props
}: Props) => {
  return (
    <Container {...props}>
      <Icon color={color} fillPercentage={fillPercentage} />
    </Container>
  );
};
