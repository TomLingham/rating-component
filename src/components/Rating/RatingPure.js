// @flow

import * as React from 'react';
import styled from 'styled-components';

import RatingItem from './RatingItem';
import { type RatingIconType } from '../RatingIcons';

const RatingItemsContainer = styled.div`
  display: flex;
  width: 80%;
  margin: 0 auto;
`;

type Props = {
  average: number,
  color: string,
  icon: RatingIconType,
  indicativeRating: number,
  max: number,
  rating: number,
  onChangeIndicativeRating?: number => void,
  onChangeRating?: number => void,
  onSelectRating?: number => void,
};

const noop = () => {};

export default ({
  average,
  color,
  icon,
  indicativeRating,
  max,
  rating,

  // we can default these callbacks to noop so that they don't error if they
  // aren't provided. Documentation is the best place for this behaviour.
  onChangeIndicativeRating = noop,
  onChangeRating = noop,
  onSelectRating = noop,
}: Props) => {
  const ratingItems = [];

  // TODO: Improve this
  // This could put the index in a data attribute, then we could pass through a
  // generic function that reads the attribute instead of creating a new
  // function for each rating item, but it's possibly a premature optimisation
  // in this case.
  for (let ratingItemIndex = 1; ratingItemIndex <= max; ratingItemIndex++) {
    const isSelected = rating >= ratingItemIndex;
    const isActive = indicativeRating >= ratingItemIndex;
    let fillPercentage = 1;

    if (!isActive && !isSelected) {
      // Bound the values here too just to make the api a little nicer to test
      fillPercentage = Math.max(Math.min(average - ratingItemIndex + 1, 1), 0);
    }

    ratingItems.push(
      <RatingItem
        key={ratingItemIndex}
        active={isActive}
        selected={isSelected}
        fillPercentage={fillPercentage}
        color={color}
        icon={icon}
        onClick={() => onChangeRating(ratingItemIndex)}
        onMouseEnter={() => onChangeIndicativeRating(ratingItemIndex)}
      />
    );
  }

  return (
    <RatingItemsContainer onMouseLeave={() => onChangeIndicativeRating(0)}>
      {ratingItems}
    </RatingItemsContainer>
  );
};
