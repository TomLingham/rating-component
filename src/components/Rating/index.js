// @flow

import * as React from 'react';

import RatingPure from './RatingPure';
import { type RatingIconType } from '../RatingIcons';

type State = {
  rating: number,
  indicativeRating: number,
};

// This is the public API for this component.
// TODO: add/generate docs
type Props = {
  average?: number,
  max?: number,
  color?: string,

  // Icon is compulsory so that we don't automatically import an icon that the
  // user does not need so that we can eventually take advantage of tree
  // shaking.
  icon: RatingIconType,
  onChangeRating: number => void,
};

export default class Rating extends React.Component<Props, State> {
  static defaultProps = {
    // We can noop this instead of erroring for a better dev experience.
    onChangeRating: () => {},
  };

  state = {
    rating: 0,
    indicativeRating: 0,
  };

  onChangeIndicativeRating = (indicativeRating: number) => {
    this.setState({ indicativeRating });
  };

  onSelectRating = (rating: number) => {
    this.setState({ rating }, () => this.props.onChangeRating(rating));
  };

  render() {
    const { icon, max = 5, average = 0, color = '#ffd60b' } = this.props;

    if (__DEV__) {
      if (!icon) throw new Error('`icon` prop must be provided.');
      if (max < 1) throw new Error('`max` prop must be greater than or equal to 1.');
      if (average < 0) throw new Error('`average` prop must be greater than or equal to 0.');
    }

    return (
      <RatingPure
        icon={icon}
        color={color}
        max={max}
        average={average}
        rating={this.state.rating}
        indicativeRating={this.state.indicativeRating}
        onChangeIndicativeRating={this.onChangeIndicativeRating}
        onChangeRating={this.onSelectRating}
      />
    );
  }
}
