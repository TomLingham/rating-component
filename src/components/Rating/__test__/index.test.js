// @flow

import * as React from 'react';
import renderer from 'react-test-renderer';
import { shallow, mount, render } from 'enzyme';

import Rating from '../';
import RatingPure from '../';
import RatingItem from '../RatingItem';

const MockIcon = () => <div>+</div>;

describe('<Rating />', () => {
  it('should render the default snapshot', () => {
    const tree = renderer.create(<Rating icon={MockIcon} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should have the same number rating items as the value passed in the prop "max"', () => {
    // Dive down so that the RatingPure component is expanded.
    const comp = shallow(<Rating max={8} icon={MockIcon} />).dive();

    expect(comp.find(RatingItem)).toHaveLength(8);
  });

  it('should call onChangeRating with the rating value that was selected', () => {
    const onChangeRating = jest.fn();
    const comp = shallow(
      <Rating max={8} onChangeRating={onChangeRating} icon={MockIcon} />
    ).dive();

    comp.childAt(2).simulate('click');

    expect(onChangeRating).toHaveBeenCalledWith(3);
  });

  it('should display the average rating that was passed in.', () => {
    const comp = shallow(
      <Rating max={8} average={5.3} icon={MockIcon} />
    ).dive();

    // Loop over all the rating items. `fillPercentage` should add up to the
    // average rating.
    const result = comp
      .find(RatingItem)
      .reduce((avg, next) => avg + next.props().fillPercentage, 0);

    expect(result).toBe(5.3);
  });

  it('should throw if no `icon` prop is passed.', () => {
    // $FlowFixMe : supress Flow error.
    expect(() => shallow(<Rating />)).toThrowErrorMatchingSnapshot();
  });

  it('should throw if `max` prop is less than 1.', () => {
    expect(() =>
      shallow(<Rating icon={MockIcon} max={0} />)
    ).toThrowErrorMatchingSnapshot();
  });

  it('should throw if `average` prop is less than 0.', () => {
    expect(() =>
      shallow(<Rating icon={MockIcon} average={-1} />)
    ).toThrowErrorMatchingSnapshot();
  });
});
