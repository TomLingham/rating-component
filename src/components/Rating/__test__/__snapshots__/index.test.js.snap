// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Rating /> should render the default snapshot 1`] = `
.c1 {
  cursor: pointer;
  padding: 0 4px;
  width: 100%;
  -webkit-transition: zoom 180ms,opacity 180ms;
  transition: zoom 180ms,opacity 180ms;
  opacity: 0.5;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 80%;
  margin: 0 auto;
}

<div
  className="c0"
  onMouseLeave={[Function]}
>
  <span
    className="c1"
    onClick={[Function]}
    onMouseEnter={[Function]}
    selected={false}
  >
    <div>
      +
    </div>
  </span>
  <span
    className="c1"
    onClick={[Function]}
    onMouseEnter={[Function]}
    selected={false}
  >
    <div>
      +
    </div>
  </span>
  <span
    className="c1"
    onClick={[Function]}
    onMouseEnter={[Function]}
    selected={false}
  >
    <div>
      +
    </div>
  </span>
  <span
    className="c1"
    onClick={[Function]}
    onMouseEnter={[Function]}
    selected={false}
  >
    <div>
      +
    </div>
  </span>
  <span
    className="c1"
    onClick={[Function]}
    onMouseEnter={[Function]}
    selected={false}
  >
    <div>
      +
    </div>
  </span>
</div>
`;

exports[`<Rating /> should throw if \`average\` prop is less than 0. 1`] = `"\`average\` prop must be greater than or equal to 0."`;

exports[`<Rating /> should throw if \`max\` prop is less than 1. 1`] = `"\`max\` prop must be greater than or equal to 1."`;

exports[`<Rating /> should throw if no \`icon\` prop is passed. 1`] = `"\`icon\` prop must be provided."`;
